emily/lambda/src/operations/chainstate.rs

```diff
@@ -2,50 +2,148 @@
+        .set_item(Some(item))
+        .send()
+        .await
+        // TODO:
+        // Make error provide less internal info.
+        .map_err(|e|
+            EmilyApiError::InternalService(
+                format!("Error occurred for table {:?} {:?}", context.settings.chainstate_table_name.clone(), e)
+            )
+        )?;
+
+    common::package_response(chainstate_entry, 201)
 }
 
 /// Handles retrieval of the current chainstate
-pub fn handle_get_chainstate(
-    _path_parameters:  HashMap<String, String>
+pub async fn handle_get_chainstate(
+    _path_parameters: HashMap<String, String>,
+    context: &LambdaContext,
 ) -> Result<common::SimpleApiResponse, errors::EmilyApiError> {
-    // path_parameters.keys()
-    Ok(models::GetChainstateResponseContent {
-        block_hash: "Tortoise".to_string(),
-        block_height: 42.0,
-    })
-    .and_then(|response| {
-        common::package_response(response, 200)
-    })
+    // // path_parameters.keys()
+    let height: &String = _path_parameters.get("height").unwrap();
+    let query_output = context.dynamodb_client
+        .query()
+        .table_name(context.settings.chainstate_table_name.clone())
+        .key_condition_expression("#h = :height")
+        .expression_attribute_names("#h", "BlockHeight")
+        .expression_attribute_values(":height", AttributeValue::N(height.clone()))
+        .send()
+        .await
+        // TODO:
+        // Make error provide less internal info.
+        .map_err(|e|
+            EmilyApiError::InternalService(
+                format!("Error occurred for table {:?} {:?}", context.settings.chainstate_table_name.clone(), e)
+            )
+        )?;
+
+    // Change behavior depending on how many entries were found.
+    match query_output.items().len() {
+
+        // No blocks found.
+        0 => Err(EmilyApiError::NotFound(
+            format!("No bock hash found for height {}", height)
+        )),
+
+        // A single block found; happy path.
+        1 => {
+            // Extract.
+            let item = query_output.items().first().unwrap();
+
+            // TODO:
+            // extract fields more safely.
+            let block_hash = item.get("BlockHash").unwrap().as_s().unwrap();
+            let block_height = item.get("BlockHeight").unwrap().as_n().unwrap();
+
+            // Package.
+            common::package_response(models::GetChainstateResponseContent {
+                block_hash: block_hash.clone(),
+                block_height: block_height.parse::<f64>().unwrap(),
+            }, 200)
+        },
+
+        // Multiple conflicting blocks found.
+        _ => {
+            let block_hashes = query_output.items()
+                .iter()
+                .map(|item| {
+                    item.get("BlockHash").unwrap().as_s().unwrap().clone()
+                })
+                .collect::<Vec<String>>();
+
+            Err(EmilyApiError::InternalService(
```

Comment:
Why should the existence of multiple conflicting blocks with the same height lead to an error? This kind of thing should be "normal", since forking would lead to this outcome right? If so, maybe we should add a unix timestamp attribute to this table and just fetch the latest one if there are multiples?

Reply:
Actually, I think that all of these tables should have a timestamp attribute for when they were created. And maybe another for when it was last updated too.